### ISSUE TITLE
Record block-level convergence diagnostics in convergence trajectory

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#209]
# Record block-level convergence diagnostics in convergence trajectory

## Summary

Extend the `convergence_trajectory_df` returned by `jaxmodels.fit()` to capture all per-block optimization diagnostics (error, stepsize, inner iterations), per-condition parameter snapshots (α, θ, β0, sparsity), and a per-variant normalized loss — data that is currently only printed in verbose mode and discarded.

## Background

The fitting process in `multidms` uses block coordinate descent with four optimization blocks per outer iteration: calibration, β0, β_nonbundle, and β_bundle. Each block's jaxopt optimizer produces diagnostic state (error, stepsize, inner iteration count) that is valuable for evaluating convergence quality.

**What currently exists:**
- `convergence_trajectory_df` records only 4 columns per iteration: `iteration`, `objective_total_trajectory`, `objective_error_trajectory`, `loss_trajectory`.
- Verbose mode (`verbose=True`) prints block-level error/stepsize/iter_num, per-condition α/θ/β0 values, and sparsity diagnostics — but this information is not stored anywhere.
- `ModelCollection.fit_models()` runs with `verbose=False` by default, meaning block-level diagnostics are never available for batch-fitted models.

**What is missing:**
- No programmatic access to block-level convergence behavior for post-hoc analysis.
- No way to diagnose which block is slow to converge, or to compare convergence profiles across models in a collection.
- No per-variant normalized loss, making it hard to compare loss across datasets of different sizes.

## Proposed Approach

Add new columns to the existing `convergence_trajectory_df` DataFrame that capture all information currently only printed in verbose mode. The data is already computed during fitting (jaxopt state objects have `.error`, `.stepsize`, `.iter_num` as attributes); we just need to store it instead of discarding it.

**Key design decisions:**
- **Single tidy DataFrame** rather than a separate structure — keeps the API surface unchanged and makes downstream analysis straightforward.
- **Always record, regardless of `verbose`** — the state attributes are already computed by jaxopt during `opt.run()`, so extracting them adds negligible overhead (just Python dict insertions of scalar values). The sparsity calculation is a lightweight array comparison. The only new computation is the per-variant normalized loss (one extra division).
- **Per-condition columns use a `{condition_name}` naming convention** — e.g., `alpha_{condition}`, `beta0_{condition}`, `sparsity_{condition}` — to keep the DataFrame tidy with one row per iteration.

### Programmatic API

```python
# Existing usage — unchanged
model = multidms.Model(data, ...)
model.fit()
df = model.convergence_trajectory_df

# New columns available on the same DataFrame
df.columns
# Index(['iteration',
#        'objective_total_trajectory', 'objective_error_trajectory',
#        'loss_trajectory', 'loss_per_variant_trajectory',
#        'calibration_error', 'calibration_stepsize', 'calibration_iter_num',
#        'beta0_error', 'beta0_stepsize', 'beta0_iter_num',
#        'beta_nonbundle_error', 'beta_nonbundle_stepsize', 'beta_nonbundle_iter_num',
#        'beta_bundle_error', 'beta_bundle_stepsize', 'beta_bundle_iter_num',
#        'alpha_condA', 'alpha_condB', ...,
#        'theta_condA', 'theta_condB', ...,
#        'beta0_condA', 'beta0_condB', ...,
#        'sparsity_condA', 'sparsity_condB', ...],
#       dtype='object')

# Example: check which block was slowest to converge
df[['calibration_error', 'beta0_error',
    'beta_nonbundle_error', 'beta_bundle_error']].plot()

# Example: compare normalized loss across models in a collection
for _, row in model_collection.models.iterrows():
    m = row['model']
    traj = m.convergence_trajectory_df
    traj['loss_per_variant_trajectory'].plot(label=str(row.name))
```

### 1. `jaxmodels.py` — Extend `trajectory_rows` dict

**File:** `multidms/jaxmodels.py` (existing)

Modify the trajectory recording block (~lines 718–728) to include all block-level state and per-condition values. Move the sparsity and per-condition parameter extraction outside the `if verbose:` guard.

```python
# Compute total training set size for normalization
n_variants_total = sum(
    data_sets[d].functional_scores.shape[0] for d in data_sets
)

# ... inside the loop, after all blocks and objective computation ...

# Per-condition diagnostics (moved outside verbose guard)
per_condition = {}
for d in model.φ:
    per_condition[f"alpha_{d}"] = float(model.α[d])
    per_condition[f"theta_{d}"] = float(jnp.exp(model.logθ[d]))
    per_condition[f"beta0_{d}"] = float(model.φ[d].β0)
    if d != model.reference_condition:
        sparsity = float(
            (model.φ[d].β - model.φ[model.reference_condition].β == 0).mean()
        )
        per_condition[f"sparsity_{d}"] = sparsity

loss_total = float(sum(loss_fn(model, data_sets, **loss_kwargs).values()))

trajectory_rows.append(
    {
        "iteration": k,
        "objective_total_trajectory": float(obj * scale),
        "objective_error_trajectory": float(objective_error),
        "loss_trajectory": loss_total,
        "loss_per_variant_trajectory": loss_total / n_variants_total,
        # block-level diagnostics
        "calibration_error": float(state_calibration.error),
        "calibration_stepsize": float(state_calibration.stepsize),
        "calibration_iter_num": int(state_calibration.iter_num),
        "beta0_error": float(state_beta0.error),
        "beta0_stepsize": float(state_beta0.stepsize),
        "beta0_iter_num": int(state_beta0.iter_num),
        "beta_nonbundle_error": float(state_nonbundle.error),
        "beta_nonbundle_stepsize": float(state_nonbundle.stepsize),
        "beta_nonbundle_iter_num": int(state_nonbundle.iter_num),
        "beta_bundle_error": float(state_bundle.error),
        "beta_bundle_stepsize": float(state_bundle.stepsize),
        "beta_bundle_iter_num": int(state_bundle.iter_num),
        # per-condition parameters
        **per_condition,
    }
)
```

### 2. `jaxmodels.py` — Update empty DataFrame columns

**File:** `multidms/jaxmodels.py` (existing)

The fallback empty DataFrame (~lines 742–750) needs to list all new column names. Best approach: define the column list once at the top of the function body based on the condition names, and reuse it for both the empty case and as a sanity check.

```python
# Define expected columns from condition names
conditions = list(data_sets.keys())
base_columns = [
    "iteration",
    "objective_total_trajectory",
    "objective_error_trajectory",
    "loss_trajectory",
    "loss_per_variant_trajectory",
    "calibration_error", "calibration_stepsize", "calibration_iter_num",
    "beta0_error", "beta0_stepsize", "beta0_iter_num",
    "beta_nonbundle_error", "beta_nonbundle_stepsize", "beta_nonbundle_iter_num",
    "beta_bundle_error", "beta_bundle_stepsize", "beta_bundle_iter_num",
]
condition_columns = []
for d in conditions:
    condition_columns.extend([f"alpha_{d}", f"theta_{d}", f"beta0_{d}"])
    if d != reference_condition:
        condition_columns.append(f"sparsity_{d}")
all_columns = base_columns + condition_columns
```

### 3. `model.py` — No changes needed

The wrapper `Model` class already stores and exposes `_convergence_trajectory_df` as-is. The new columns will be available automatically through the existing `convergence_trajectory_df` property.

### 4. `model_collection.py` — No changes needed

`ModelCollection` already stores models with their trajectory DataFrames. The richer data will be available on each model's `convergence_trajectory_df`.

### Overhead analysis

- **Block state attributes** (`error`, `stepsize`, `iter_num`): These are already computed by jaxopt's `run()` method and stored on the returned state object. Reading them is a simple attribute access — zero additional computation.
- **Per-condition α, θ, β0**: These are already stored on the model — just `float()` conversion of existing JAX scalars.
- **Sparsity**: One elementwise comparison + mean over the β vector. Negligible relative to the gradient computations in each block.
- **`loss_per_variant_trajectory`**: One floating-point division of an already-computed value. The `n_variants_total` sum is computed once before the loop.
- **Memory**: ~20–30 extra scalar values per iteration in a Python dict. For typical fits (10–50 iterations), this is a few KB.

### Per-variant normalized loss

The `loss_per_variant_trajectory` column divides the total loss (sum across all conditions) by the total number of variants across all conditions. This enables fair comparison of convergence behavior across datasets of different sizes. The denominator is:

```python
n_variants_total = sum(d.functional_scores.shape[0] for d in data_sets.values())
```

This is computed once before the fitting loop begins (constant across iterations).

### Backward compatibility

- The DataFrame gains new columns but retains all existing ones with identical values.
- Code that accesses `convergence_trajectory_df` by column name (e.g., `df["loss_trajectory"]`) will work unchanged.
- Code that unpacks all columns positionally (unlikely given the API) would need updating.
- Pickled `ModelCollection` objects from before this change will have the old DataFrame schema. This is already the case for any schema change and is acceptable.

## Testing Strategy

- **Unit test for new columns**: Fit a small model and assert all expected columns are present in `convergence_trajectory_df`.
- **Unit test for values**: Check that block-level error values are non-negative, stepsize > 0, iter_num >= 0.
- **Unit test for per-variant normalization**: Manually compute `loss_trajectory / n_variants` and assert it matches `loss_per_variant_trajectory`.
- **Unit test for per-condition columns**: Verify that condition-specific columns (α, θ, β0, sparsity) are present and named correctly.
- **Regression test**: Verify that existing columns (`iteration`, `objective_total_trajectory`, `objective_error_trajectory`, `loss_trajectory`) have identical values to the current implementation.
- Edge cases:
  - Single-condition model (no sparsity columns for reference condition)
  - Model that converges in 1 iteration
  - KeyboardInterrupt during fitting (existing behavior preserved)
  - Empty trajectory (0 iterations)

## Documentation Updates

- [ ] Docstring for `jaxmodels.fit()` — update the return value description to list new columns
- [ ] Docstring for `Model.convergence_trajectory_df` property — mention new columns
- [ ] Example notebook showing block-level convergence analysis (future PR)

## Dependencies

None — all new functionality uses existing jaxopt state attributes and standard Python/pandas operations.

### Alternative 1: Separate `block_trajectory_df` attribute

Store block-level data in a second DataFrame with a different schema (one row per block per iteration, i.e., 4× more rows).

**Why not chosen:** Adds API surface area for something that fits naturally as extra columns on the existing DataFrame. A single DataFrame is easier to filter, join, and plot.

### Alternative 2: Structured logging / callback

Add a callback system that users can hook into for custom logging during fitting.

**Why not chosen:** Over-engineered for the current need. The data is small and well-defined. A callback system could be a future enhancement if more flexible monitoring is needed.

### Alternative 3: Only record when `verbose=True`

Keep the current pattern of only computing diagnostics in verbose mode.

**Why not chosen:** Users frequently want to analyze convergence post-hoc, especially in `ModelCollection` batch fits where verbose output is impractical. The overhead is negligible.

## Success Criteria

- [ ] `convergence_trajectory_df` contains all block-level diagnostics (error, stepsize, iter_num) for all 4 blocks
- [ ] Per-condition α, θ, β0, and sparsity columns present and correctly named
- [ ] `loss_per_variant_trajectory` column present and correctly normalized
- [ ] All diagnostics recorded regardless of `verbose` setting
- [ ] Existing columns unchanged (backward compatible)
- [ ] All existing tests pass
- [ ] New unit tests for column presence, value correctness, and edge cases

## Future Work

- Visualization methods for block-level convergence (separate PR, as discussed)
- Per-condition loss columns (not just total) for diagnosing condition-specific fitting issues
- Callback system for custom monitoring during fitting
- Integration with experiment pipeline notebooks for convergence quality checks

## Enhancements to CLAUDE.md

- Note that `convergence_trajectory_df` now contains block-level diagnostics and per-condition parameter snapshots, not just the 3 summary columns — update the "Key design patterns" or "Data flow" section to reflect this richer convergence tracking.
